### ISSUE TITLE
How can it not affect another monitor

### ISSUE THREAD

**Issue by zdmgmail** [stoeckmann/xwallpaper#20]
When I set a wallpaper to be displayed, the wallpaper on the other monitor will be blank. Is there any way?

```
xwallpaper --output HDMI-1-1 --zoom  /path/a.png
```

**Comment by zdmgmail** [comment]
```
~ » xwallpaper --version                                                                 zdm@ar
0.6.4
------------------------------------------------------------------------------------------------
~ » uname -a                                                                             zdm@ar
Linux ar 5.6.3-arch1-1 #1 SMP PREEMPT Wed, 08 Apr 2020 07:47:16 +0000 x86_64 GNU/Linux

```

**Comment by zdmgmail** [comment]
I can use

```
xwallpaper --output $monitor1 --zoom "a.jpg --output $monitor2 --zoom b.jpg
```

but just want to set one

**Comment by stoeckmann** [comment]
Hi zdmgmail,

from a user perspective, this sounds like a trivial thing to do. From an X11 perspective, this is not easy at all.

1. X11 uses a root window which spans across the whole screen. A screen in that sense is the combination of all monitors you use (as long as you have just one x server running).
2. The root window can have a pixmap as background (i.e. a wallpaper), which tiles automatically. You can experience this if you set a wallpaper when your monitor is set to 800x600 and then you increase it 1024x768 and so on.
3. To have the visual experience of independent wallpapers on multiple monitors, programs like xwallpaper have to retrieve the exact positions of monitors on this one large pixmap and draw wallpapers into correct positions. Like, composing a large picture from small ones. Eventually, this one large pixmap which spans across the whole root window is uploaded.
4. Programs cannot retrieve the background image of the root window. Semi-transparent terminals like urxvt or aterm etc. use a pre-defined atom. Let's say it's some kind of shared storage within the X server. If the atom is not set, semi-transparent terminals do not work.

Programs like nitrogen are able to set a wallpaper on a single monitor (maybe it fits your needs). But what they do internally is:
- Retrieve content of atom. If it's a suitable pixmap, use that for future operations
- Draw into the pixmap
- Update background of X root window and update atom

I could implement that in xwallpaper as well. After all, xwallpaper supports atoms for semi-transparent terminals as well. But...

If you ask me, it's an incomplete solution. It only works if the atom has been set before. If it's not set, you have to fall back to current solution. If you have set your root window to a solid color with xsetroot, you couldn't retrieve that either. You cannot even figure out if your X server has a zig-zag pattern or a black background.

So right now I have a tendency to do it right or not at all, and due to X11's nature, a right solution is basically not available.

May I ask for which kind of setup you want to use xwallpaper? Perhaps I can get a better grasp of how to implement it "as good as possible" or recommend a tool which fits your needs ([nitrogen](https://github.com/l3ib/nitrogen/) is an option).

**Comment by zdmgmail** [comment]
> Hi zdmgmail,
> 
> from a user perspective, this sounds like a trivial thing to do. From an X11 perspective, this is not easy at all.
> 
> 1. X11 uses a root window which spans across the whole screen. A screen in that sense is the combination of all monitors you use (as long as you have just one x server running).
> 2. The root window can have a pixmap as background (i.e. a wallpaper), which tiles automatically. You can experience this if you set a wallpaper when your monitor is set to 800x600 and then you increase it 1024x768 and so on.
> 3. To have the visual experience of independent wallpapers on multiple monitors, programs like xwallpaper have to retrieve the exact positions of monitors on this one large pixmap and draw wallpapers into correct positions. Like, composing a large picture from small ones. Eventually, this one large pixmap which spans across the whole root window is uploaded.
> 4. Programs cannot retrieve the background image of the root window. Semi-transparent terminals like urxvt or aterm etc. use a pre-defined atom. Let's say it's some kind of shared storage within the X server. If the atom is not set, semi-transparent terminals do not work.
> 
> Programs like nitrogen are able to set a wallpaper on a single monitor (maybe it fits your needs). But what they do internally is:
> 
> * Retrieve content of atom. If it's a suitable pixmap, use that for future operations
> * Draw into the pixmap
> * Update background of X root window and update atom
> 
> I could implement that in xwallpaper as well. After all, xwallpaper supports atoms for semi-transparent terminals as well. But...
> 
> If you ask me, it's an incomplete solution. It only works if the atom has been set before. If it's not set, you have to fall back to current solution. If you have set your root window to a solid color with xsetroot, you couldn't retrieve that either. You cannot even figure out if your X server has a zig-zag pattern or a black background.
> 
> So right now I have a tendency to do it right or not at all, and due to X11's nature, a right solution is basically not available.
> 
> May I ask for which kind of setup you want to use xwallpaper? Perhaps I can get a better grasp of how to implement it "as good as possible" or recommend a tool which fits your needs ([nitrogen](https://github.com/l3ib/nitrogen/) is an option).

Thank you for your very clear and detailed reply. I never thought of these and  very happy to learn about them.

 I have a script, press the shortcut key to randomly set the wallpaper of several monitors

 But sometimes I use sxiv to browse pictures. I want to press the shortcut key to execute a script to set this picture as the wallpaper of a certain monitor.(prevent set 1080x1920 for 1920x1080)

But this demand is not big, may filter the wallpaper first, and then always set it randomly by different directory


the script 

```
monitor1=DP-1
monitor2=HDMI-1-1

bgLog="$HOME/log/bg.log"

bgloc="${XDG_CACHE_HOME:-$HOME/.cache/}/bg"
bgloc2="${XDG_CACHE_HOME:-$HOME/.cache/}/bg"

if [[ -f "$1" ]]; then
	ln -sf "$(readlink -f "$1")" "$bgloc"
	name=$(find "$(readlink -f "$1")" -iregex '.*.\(jpg\|jpeg\|png\|gif\)' -type f | shuf -n 1)
	xwallpaper --zoom "$bgloc"
	exit
fi

dir1=$HOME/GitHub/Wallpapers/Vertical/
name=$(find "$(readlink -f "$dir1")" -type f | shuf -n 1)
bgloc=$name

dir2=$HOME/GitHub/Wallpapers/Horizontal/
name=$(find "$(readlink -f "$dir2")" -type f | shuf -n 1)
bgloc2=$name

xwallpaper --output $monitor1 --zoom "$bgloc" --output $monitor2 --zoom "$bgloc2"

echo "$bgloc"  >> $bgLog
echo "$bgloc2" >> $bgLog

```

**Comment by stoeckmann** [comment]
I see, so it's not like xwallpaper is run in parallel with another tool which is supposed to handle a single monitor.

In that case I recommend to adjust your script slightly:
- Keep symbolic links to currently selected wallpapers in "bgLoc"
- Always call xwallpaper at the end with both outputs specified
- If no file is specified, set symbolic links to random wallpapers
- If a file is specified, replace a single symbolic link

This way, it will look like only one monitor is updated, even though xwallpaper redraws the whole screen.

```
monitor1=DP-1
monitor2=HDMI-1-1

bgLoc="${XDG_CACHE_HOME:-$HOME/.cache}/bg"

random_wallpaper() {
	dir=$1
	monitor=$2
	name=$(find "$(readlink -f "$dir")" -type f | shuf -n 1)
	ln -sf "$name" "$bgLoc/$monitor"
}

if [[ -f "$1" ]]; then
	# TODO: check which monitor applies best
	ln -sf "$(readlink -f "$1")" "$bgLoc/$monitor1"
else
	random_wallpaper "$HOME/GitHub/Wallpapers/Vertical/" "$monitor1"
	random_wallpaper "$HOME/GitHub/Wallpapers/Horizontal/" "$monitor2"
fi

xwallpaper \
	--output "$monitor1" --zoom "$bgLoc/$monitor1" \
	--output "$monitor2" --zoom "$bgLoc/$monitor2"
```

**Comment by zdmgmail** [comment]
> I see, so it's not like xwallpaper is run in parallel with another tool which is supposed to handle a single monitor.
> 
> In that case I recommend to adjust your script slightly:
> 
> * Keep symbolic links to currently selected wallpapers in "bgLoc"
> * Always call xwallpaper at the end with both outputs specified
> * If no file is specified, set symbolic links to random wallpapers
> * If a file is specified, replace a single symbolic link
> 
> This way, it will look like only one monitor is updated, even though xwallpaper redraws the whole screen.
> 
> ```
> monitor1=DP-1
> monitor2=HDMI-1-1
> 
> bgLoc="${XDG_CACHE_HOME:-$HOME/.cache}/bg"
> 
> random_wallpaper() {
> 	dir=$1
> 	monitor=$2
> 	name=$(find "$(readlink -f "$dir")" -type f | shuf -n 1)
> 	ln -sf "$name" "$bgLoc/$monitor"
> }
> 
> if [[ -f "$1" ]]; then
> 	# TODO: check which monitor applies best
> 	ln -sf "$(readlink -f "$1")" "$bgLoc/$monitor1"
> else
> 	random_wallpaper "$HOME/GitHub/Wallpapers/Vertical/" "$monitor1"
> 	random_wallpaper "$HOME/GitHub/Wallpapers/Horizontal/" "$monitor2"
> fi
> 
> xwallpaper \
> 	--output "$monitor1" --zoom "$bgLoc/$monitor1" \
> 	--output "$monitor2" --zoom "$bgLoc/$monitor2"
> ```

I did and 

``` sh

ln: failed to create symbolic link '/home/zdm/.cache/bg/DP-1': No such file or directory
ln: failed to create symbolic link '/home/zdm/.cache/bg/HDMI-1-1': No such file or directory

so 
mkdir -p "$bgLoc"

and use 

if [[ -f "$1" ]]; then
	# TODO: check which monitor applies best
	[[ "$2" == "1" ]] \
		&& ln -sf "$(readlink -f "$1")" "$bgLoc/$monitor1" \
		|| ln -sf "$(readlink -f "$1")" "$bgLoc/$monitor2"

then add in sxiv 

        $"w") ~/.config/short_script/setbg "$file" 1 & ;;
        $"W") ~/.config/short_script/setbg "$file" 2 & ;;

``

It works very fine !!

whole script

```sh

monitor1=DP-1
monitor2=HDMI-1-1

bgLoc="${XDG_CACHE_HOME:-$HOME/.cache}/bg"
mkdir -p "$bgLoc"

random_wallpaper() {
	dir=$1
	monitor=$2
	name=$(find "$(readlink -f "$dir")" -type f | shuf -n 1)
	ln -sf "$name" "$bgLoc/$monitor"
}

if [[ -f "$1" ]]; then
	# TODO: check which monitor applies best
	[[ "$2" == "1" ]] \
		&& ln -sf "$(readlink -f "$1")" "$bgLoc/$monitor1" \
		|| ln -sf "$(readlink -f "$1")" "$bgLoc/$monitor2"
else
	random_wallpaper "$HOME/GitHub/Wallpapers/Vertical/" "$monitor1"
	random_wallpaper "$HOME/GitHub/Wallpapers/Horizontal/" "$monitor2"
fi

xwallpaper \
	--output "$monitor1" --zoom "$bgLoc/$monitor1" \
	--output "$monitor2" --zoom "$bgLoc/$monitor2"
```

**Comment by zdmgmail** [comment]
It works very fine !!

**Comment by stoeckmann** [comment]
Great to hear. :)
Can you close the issue if it works for you? I would prefer to not adjust the xwallpaper code due to the arguments in my first reply.

**Comment by stoeckmann** [comment]
Hi again @zdmgmail and @mjkalyan.

I gave these topics more thoughts during the weekend. My arguments are still in place, but I have adjusted the code. As long as it's clear in the documentation, nobody should be surprised about these weird effects. For example setting a solid color with xsetroot and then setting a single output with xwallpaper.

I have taken into consideration that simply setting a single output is more of a surprise than blanking the screen on initial xwallpaper run.

Talking about it, I did not find an available program which allows me to reset atom contents. Or simply resetting an already set wallpaper. I have added two more options (--no-root and --no-atoms) which allows strict separation between modifying the "read-only" X root window background and the "read and write" atoms.

It allows quite some cool features like having different wallpapers on both layers. It leads to the effect that pseudo transparent programs like rxvt-unicode show a different image. An example would be: A sunny landscape on your desktop but the same landscape at night in your terminal emulator.

Let me know what you think about it and especially if manual page is precise enough.
